### PR TITLE
add conversion from arrow to tensor

### PR DESCRIPTION
**Pull request by poopoothegorilla**
This PR adds the ability to convert an Arrow Array to a Dense Tensor. 

Notes:
- [x] Add Tests
- [x] Adds Basic convesions for `FromArrowArray` function to Dense Tensor
- [x] Adds ArrowToTensor Example
- [x] Updates the Native Iterators to work with newer versions of Golang

### CODE REVIEW

File: dense_compat.go [hunk 1]
```diff
@@ -433,3 +435,79 @@ func ToMat64(t *Dense, opts ...FuncOpt) (retVal *mat.Dense, err error) {
...
+	// TODO(poopoothegorilla): instead of creating bool ValidMask maybe
+	// bitmapBytes can be used from arrow API
+	mask := make([]bool, r)
+	for i := 0; i < r; i++ {
+		mask[i] = a.IsNull(i)
+	}
+
+	switch a.DataType() {
+	case arrow.BinaryTypes.String:
+		backing := make([]string, a.Len())
```
Comment (poopoothegorilla) [inline dense_compat.go:455]: use r instead

File: genlib2/dense_compat.go [hunk 2]
```diff
@@ -237,17 +243,91 @@ func ToMat64(t *Dense, opts ...FuncOpt) (retVal *mat.Dense, err error) {
...
 func init() {
+	importsArrow = template.Must(template.New("imports_arrow").Funcs(funcs).Parse(importsArrowRaw))
 	conversions = template.Must(template.New("conversions").Funcs(funcs).Parse(conversionsRaw))
 	compats = template.Must(template.New("compat").Funcs(funcs).Parse(compatRaw))
+	compatsArrow = template.Must(template.New("compat_arrow").Funcs(funcs).Parse(compatArrowRaw))
 }
 
 func generateDenseCompat(f io.Writer, generic Kinds) {
+	// NOTE(poopoothegorilla): an alias is needed for the Arrow Array pkg to prevent naming
+	// colisions
```
Comment (poopoothegorilla) [inline genlib2/dense_compat.go:323]: colisions -> collisions

File: genlib2/native_iterator.go [hunk 3]
```diff
@@ -54,17 +56,16 @@ func Matrix{{short .}}(t *Dense) (retVal [][]{{asType .}}, err error) {
...
+		hdr := (*reflect.SliceHeader)(unsafe.Pointer(&retVal[i]))
+		hdr.Data = uintptr(unsafe.Pointer(&data[start]))
+		hdr.Cap = cols
+		hdr.Len = cols
 	}
 	return
 }
 
-// Tensor3{{short .}} converts a *Dense into  a [][][]{{asType .}}. 
+// Tensor3{{short .}} converts a *Dense into a [][][]{{asType .}}.
```
Comment (poopoothegorilla) [inline genlib2/native_iterator.go:68]: formating issue

File: genlib2/native_iterator.go [hunk 4]
```diff
@@ -84,18 +85,16 @@ func Tensor3{{short .}}(t *Dense) (retVal [][][]{{asType .}}, err error) {
...
+			hdr := (*reflect.SliceHeader)(unsafe.Pointer(&retVal[i][j]))
+			hdr.Data = uintptr(unsafe.Pointer(&data[start]))
+			hdr.Cap = cols
+			hdr.Len = cols
 		}
 	}
 	return
-}
-`
+}`
```
Comment (poopoothegorilla) [inline genlib2/native_iterator.go:97]: formatting change

**Comment by poopoothegorilla** [comment]
I think a separate PR might be appropriate for adding Multidimensional support for Arrow. These are the tasks

- [ ] Add ListArray Arrow Datatype to allow for multi dimensional conversions
- [ ] Add Ability to create Multidimensional arrays without using the ListArray Arrow Datatype 

I think it makes sensor to do a separate PR for TensorToArrow as well

**Comment by cfgt** [comment]
@poopoothegorilla Is this PR complete by itself? If so, I'll merge it.
@chewxy is off the grid at the moment.

File: dense_compat_test.go [hunk 5]
```diff
@@ -105,3 +108,216 @@ func TestFromMat64(t *testing.T) {
...
+
+func TestFromArrowArray(t *testing.T) {
+	assert := assert.New(t)
+	var T *Dense
+	pool := memory.NewGoAllocator()
+
+	for i, taat := range toArrowArrayTests {
+		var m arrowArray.Interface
+
+		switch taat.dt {
```
Comment (poopoothegorilla) [inline dense_compat_test.go:188]: forgot to add fixed width and binary types

**Comment by poopoothegorilla** [comment]
@cfgt should be good now.

**Pull request merged**